### PR TITLE
Sc 76 add final keyword

### PR DESCRIPTION
**Pull request by VladAver87**

### CODE REVIEW

File: src/main/scala/com/blackmorse/spark/clickhouse/package.scala [hunk 1]
```diff
@@ -8,8 +8,11 @@ package object clickhouse {
   val TABLE = "TABLE"
   val BATCH_SIZE = "BATCH_SIZE"
   val CLUSTER = "CLUSTER"
+  val READ_DIRECTLY_FROM_DISTRIBUTED_TABLE = "read_directly_from_distributed"
```
Comment (Blackmorse) [inline src/main/scala/com/blackmorse/spark/clickhouse/package.scala]: Duplicate?

File: src/main/scala/com/blackmorse/spark/clickhouse/package.scala [hunk 2]
```diff
@@ -8,8 +8,11 @@ package object clickhouse {
...
   val DIRECTLY_USE_DISTRIBUTED_TABLE = "read_directly_from_distributed"
   val RANDOM_WRITES_SHUFFLE = "random_writes_shuffle"
+  val USE_FORCE_COLLAPSING_MODIFIER = "use_force_collapsing_modifier"
```
Comment (Blackmorse) [inline src/main/scala/com/blackmorse/spark/clickhouse/package.scala]: I believe `FORCE_COLLAPSING_MODIFIER` is enough.

File: src/main/scala/com/blackmorse/spark/clickhouse/utils/SqlUtils.scala [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package com.blackmorse.spark.clickhouse.utils
+
+import com.blackmorse.spark.clickhouse.tables.ClickhouseTable
+
+object SqlUtils {
+
+  def prepareSqlQuery(
```
Comment (Blackmorse) [inline src/main/scala/com/blackmorse/spark/clickhouse/utils/SqlUtils.scala:7]: In main branch Factories and merged together, so we can move this method there

File: src/main/scala/com/blackmorse/spark/clickhouse/utils/SqlUtils.scala [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+                       fields: String,
+                       table: ClickhouseTable,
+                       useForceCollapsingModifier: Boolean
+                     ): String = {
+    val sql = s"SELECT $fields FROM $table"
+    table.engine match {
+      case engine if useForceCollapsingModifier &&
```
Comment (Blackmorse) [inline src/main/scala/com/blackmorse/spark/clickhouse/utils/SqlUtils.scala]: I see no reason for using patter-matching. Probably just if-else will work

File: src/it/scala/com/blackmorse/spark/clickhouse/cluster/ClusterShardedReaderTest.scala [hunk 5]
```diff
@@ -22,6 +22,108 @@ class ClusterShardedReaderTest extends AnyFlatSpec with Matchers with DataFrameS
     shardData1 ++ shardData2
   }
 
+  private def writeDuplicateData: Seq[(Int, String)] = {
```
Comment (Blackmorse) [inline src/it/scala/com/blackmorse/spark/clickhouse/cluster/ClusterShardedReaderTest.scala]: Probably we should proceed with one generic method for writing the data? 

Signature like this:

```
def writeData[T](shard1Data: Seq[T], shard2Data: Seq[T])(implicit e: Encoder[T]): Seq[T]
```

File: src/main/scala/com/blackmorse/spark/clickhouse/utils/SqlUtils.scala (hunk 3 above)
Comment (VladAver87) [reply src/main/scala/com/blackmorse/spark/clickhouse/utils/SqlUtils.scala:7]: That's not true :)

File: src/main/scala/com/blackmorse/spark/clickhouse/package.scala (hunk 1 above)
Comment (VladAver87) [reply src/main/scala/com/blackmorse/spark/clickhouse/package.scala]: Yep)

File: src/main/scala/com/blackmorse/spark/clickhouse/package.scala (hunk 2 above)
Comment (VladAver87) [reply src/main/scala/com/blackmorse/spark/clickhouse/package.scala]: Done

File: src/main/scala/com/blackmorse/spark/clickhouse/utils/SqlUtils.scala (hunk 4 above)
Comment (VladAver87) [reply src/main/scala/com/blackmorse/spark/clickhouse/utils/SqlUtils.scala]: Done

File: src/it/scala/com/blackmorse/spark/clickhouse/cluster/ClusterShardedReaderTest.scala (hunk 5 above)
Comment (VladAver87) [reply src/it/scala/com/blackmorse/spark/clickhouse/cluster/ClusterShardedReaderTest.scala]: Done

**Review by Blackmorse — APPROVED** [review]

**Pull request merged**